sql/parser: typeCheckOverloadedExprs allocated bytes optimizations

petermattis opened the pull request:
Replace usage of indexedExpr with a slice of indexes that is passed around.
Mildly less convenient with the benefit of shrinking the associated slice
elements from 24 bytes to 8 bytes.

Use uint8 indexes for overloadedIdxs. This limits the number of overloads to 255, but cuts allocated bytes for this data structure by 1/8.

Combined, these changes reduce allocated bytes in
typeCheckOverloadedExprs by 70% (bringing it into the insignificant
realm).

petermattis committed:
sql/parser: use uint8 indexes for overloadedIdxs

This limits the number of overloads to 255, but cuts allocated bytes for
this data structure by 1/8.

petermattis committed:
sql/parser: remove indexedExpr

Replace usage of indexedExpr with a slice of indexes that is passed
around. Mildly less convenient with the benefit of shrinking the
associated slice elements from 24 bytes to 8 bytes.

See #14927

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/17152)
<!-- Reviewable:end -->

jordanlewis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>, I don't think it's too much less convenient but I would like to see some improved naming or at least some explanatory comments.

Out of curiosity, did you check the before and after of execution speed? Perhaps the memory reduction makes this change worth it regardless, but I wonder if this changed the timings perceptably.

---

Reviewed 1 of 1 files at r1, 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/sql/parser/overload.go, line 315 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17152#-Kp_RxigrVerTIoXN5CC:-Kp_RxigrVerTIoXN5CD:b-7f3vzv) ([raw file](https://github.com/cockroachdb/cockroach/blob/6d314102d1a526b6e271b762bc1b735e8885d6a9/pkg/sql/parser/overload.go#L315)):*
> ```Go
> 	for i := 0; i < len(overloads); i++ {
> 		s.overloadIdxs[i] = uint8(i)
> 	}
> ```

I have a patch that uses a bitmap for this, but it seemed to slow things down. I could share it with you if you like since you're working in this area.

---

*[pkg/sql/parser/overload.go, line 243 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17152#-Kp_Uc5g4WGlUNUeBk1w:-Kp_Uc5g4WGlUNUeBk1x:bhhjmg6) ([raw file](https://github.com/cockroachdb/cockroach/blob/44f9035d7d204d7ae5a56f698f0f71935a6a5b67/pkg/sql/parser/overload.go#L243)):*
> ```Go
> 	overloadIdxs     []uint8
> 	typedExprs       []TypedExpr
> 	resolvableExprs  []int
> ```

Please consider changing the name of these fields to say `idxs` like you did with `overloadIdxs`, or at least adding a comment explaining their relationship with `exprs`. The meaning is obvious to me now, but to a new reader it may be a bit opaque.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17152#-:-Kp_UwMyJ_6a-yBTc9hE:b-sy1lkx)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Performance is slight better on the test I ran (`kv --read-percent 100`). SQL benchmark numbers were unchanged as they don't seem to tickle this code path for some reason.

---

Review status: 2 of 3 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/sql/parser/overload.go, line 315 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17152#-Kp_RxigrVerTIoXN5CC:-Kp_bXO7ytxWqRu2iahe:bicldwi) ([raw file](https://github.com/cockroachdb/cockroach/blob/6d314102d1a526b6e271b762bc1b735e8885d6a9/pkg/sql/parser/overload.go#L315)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I have a patch that uses a bitmap for this, but it seemed to slow things down. I could share it with you if you like since you're working in this area.
</blockquote></details>

Yeah, I recall the bitmap approach. We might be able to revisit in go1.9 with the addition of `math/bits` allowing us to quickly find the first set bit.

---

*[pkg/sql/parser/overload.go, line 243 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17152#-Kp_Uc5g4WGlUNUeBk1w:-Kp_bgTNLUjE5__ipZZ9:b-xm825r) ([raw file](https://github.com/cockroachdb/cockroach/blob/44f9035d7d204d7ae5a56f698f0f71935a6a5b67/pkg/sql/parser/overload.go#L243)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Please consider changing the name of these fields to say `idxs` like you did with `overloadIdxs`, or at least adding a comment explaining their relationship with `exprs`. The meaning is obvious to me now, but to a new reader it may be a bit opaque.
</blockquote></details>

Good idea. Renamed and added a small comment.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17152#-:-Kp_cfPeb3zD0kV0qZJI:b34maoq)*
<!-- Sent from Reviewable.io -->
